installation/downloadReferences.sh

```diff
@@ -1,66 +1,148 @@
+	check_md5sum && exit 0 || echo downloading dbSNP file....
+
+	# CITATION
+	# As a NCBI Resource: "Sherry ST, Ward MH, Kholodov M, Baker J, Phan L, Smigielski EM, Sirotkin K. dbSNP: the NCBI database of genetic variation. Nucleic Acids Res. 2001 Jan 1;29(1):308-11."
+	# As a whole for a specific build (use this!) : "Database of Single Nucleotide Polymorphisms (dbSNP). Bethesda (MD): National Center for Biotechnology Information, National Library of Medicine. (dbSNP Build ID: 141 ). Available from: http://www.ncbi.nlm.nih.gov/SNP/"
+	# A single or a range of Submitted SNP (ss) or Reference SNP (rs) entries: "Database of Single Nucleotide Polymorphisms (dbSNP). Bethesda (MD): National Center for Biotechnology Information, National Library of Medicine. dbSNP accession:{ss1 or ss1 – ss100}, (dbSNP Build ID: 141). Available from: http://www.ncbi.nlm.nih.gov/SNP/"
+
+	# DOWNLOAD
+	DBSNP_BASE_URL="ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b150_GRCh37p13/VCF"
+	wget -c "$DBSNP_BASE_URL/README.txt"
+	wget -c "$DBSNP_BASE_URL/00-All.vcf.gz"
+	wget -c "$DBSNP_BASE_URL/00-All.vcf.gz.tbi"
+
+	# POST PROCESSING
+	# extract SNPs from dbSNP version 135 and older
+	zcat 00-All.vcf.gz |
+	awk '/^#/{print} /VC=SNV/{ v=$8; sub(/.*dbSNPBuildID=/, "", v); sub(/;.*/, "", v); if (v~/^[0-9]+$/ && int(v)<='$DBSNP_VERSION') print }' |
+	bgzip > 00-All.SNV.vcf.gz
+	tabix -p vcf 00-All.SNV.vcf.gz
+
+	# CLEANUP
+	rm -f 00-All.vcf.gz 00-All.vcf.gz.tbi
+
+	check_md5sum
+	)
 fi
 
-if [[  $MAPPABILITY_FILE != "true" ]]
+if [[ "$MAPPABILITY_FILE" == "true" ]]
 then
-       echo downloading mappability file....
-       wget -P databases/UCSC  http://hgdownload.soe.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeMapability/wgEncodeCrgMapabilityAlign100mer.bigWig
-       echo "please convert the bigwig file to bedgraph"
+	(
+	mkdir_cd databases/UCSC
+
+	EXPECTED_MD5SUM=3d12d0a4d7afdb52cfd10f886d48b5f0
+	check_md5sum && exit 0 || echo downloading mappability file....
+
+	wget -c http://hgdownload.soe.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeMapability/wgEncodeCrgMapabilityAlign100mer.bigWig
+	bigWigToBedGraph wgEncodeCrgMapabilityAlign100mer.bigWig /dev/stdout | bgzip > wgEncodeCrgMapabilityAlign100mer_chr.bedGraph.gz
+	tabix -p bed wgEncodeCrgMapabilityAlign100mer_chr.bedGraph.gz
+	rm -f wgEncodeCrgMapabilityAlign100mer.bigWig
+
+	check_md5sum
+	)
 fi
 
-if [[  $CHROMOSOME_LENGTH_FILE != "true" ]] 
+if [[ "$REPLICATION_TIMING_FILE" == "true" ]]
 then
-	echo downloading dbSNP file....
-       wget -qO- http://hgdownload.cse.ucsc.edu/goldenPath/hg19/database/chromInfo.txt.gz  | zcat | grep -Pv "(_)|(chrM)" | sed -e '1i\#chrom\tsize\tfileName' >stats/chrlengths.txt
+	(
+	mkdir_cd databases/ENCODE
+
+	EXPECTED_MD5SUM=2a63b34a737383af2a3f7eb32801a5fa
+	check_md5sum && exit 0 || echo downloading replication timing file....
+
+	wget -c https://raw.githubusercontent.com/eilslabs/ACEseqWorkflow/github/installation/ReplicationTime_10cellines_mean_10KB.Rda
```

Comment:
I agree with @dankwart-de that there should be less hardcoding. Often (and quite convenient) is to put all configuration variables with version numbers, URLs, MD5 sums, etc. at the top of the script. The code may already get clearer by at least putting these variables at the top of each block, instead of interleaving it with executed code. But at the top of the script is certainly preferable. Consider what a user has to do, to adapt the URLs, if we forgot it. He/She has to read the whole code, decide for every line whether to change it or not. That is much easier if all variables are at the top.

Reply:
While I usually agree that configuration variables should go to the top of a script, I don't think that it makes sense here. Changing a download URL would very likely entail the need to change additional code. For example, if we ever decide to use a different assembly (hg38 instead of hs37d5), then the name of the downloaded file would change, too. Therefore, the subsequent `gunzip` statement would fail. Similar adaptations would be necessary in the other sections.

If URLs and MD5 sums were at the top, this would mean that the user still has to read the code, in order to make sure that changes to variables don't break code. But in addition he would need to jump back and forth between the header and the body of the script. The way how the script is structured now, i.e., divided into independent sections, makes changes easier, because you can be sure that you don't have to look anywhere else in the code other than in the affected section. Plus, the sections are all just a handful of commands, so making adaptations requires understanding the few lines of code of the respective section rather than the whole code.

Reply:
* The argument the downstream code would break only holds, because it is full of hardcoded filenames as well. Why not use `$(basename "$fastaGenome")` instead of `hs37d5.fa.gz` for the unzipping? This even works with URLs instead of files. The argument does however hold in so far as the format of the files (compression, syntax, etc.) is relevant.
* Your argument of the limited scope of change (everything is local) is correct. Not sure whether it counter-weighs my argument that interspersed configuration and code is a bad idea, though -- in particular in the light of the last argument:
* Your argument that its just a few lines of code does not hold. Experience tells that even a single line of code can contain errors and may be too much for someone to digest who is completely new to the code. Most importantly, we are not talking about a one-liner here, we are talking about a total of 149 lines of code.

Don't misunderstand me. I won't go and stop you from merging the thing as is. It is just a suggestion and maybe a kind request to change the code. I really appreciate your contribution and if you think it is too much for you, leave it as it is ;-)

Reply:
> Why not use `$(basename "$fastaGenome")`

I realize that the file name was a bad example. The file format (compressed vs. non-compressed) or the need to convert bigwig to bedgraph better illustrate my concern. My point is that I am very skeptical that the code can ever be generalized enough that we can separate configuration and code, because the downloaded files will very likely entail changes in post-processing. That is why I stand by my view that we should not separate the two.

> 149 lines of code

True. Even understanding the structure of the script might take a moment. I will highlight the structure better using comment lines that clearly separate the sections at first glance.

What I was trying to say is: While all the hardcoding is ugly, I'd argue that it makes the code _easier_ to grasp/modify by someone who is new to the code, because the code reflects very naturally what someone would type on the command-line, if he wanted to download the files manually. It is an excess of abstraction and generalization of code that typically makes it hard to understand.